src/Microsoft.DotNet.Build.CloudTestTasks/UploadClient.cs

```diff
@@ -8,10 +8,12 @@
 using System.Collections.Generic;
```

Comment:
In general I'm a trying to understand the reason behind this PR. Looks like if a blob (name) already exists we check if the contents are identical? In what scenarios this is not the case? When we want to publish a package/asset that has changed but still we want to use the same version?

Reply:
I think that the overall motivation is to not consider a publish attempt a failure when overrite is false if an existing package has already been published and is identical.  ie, the package has already been published and is identical, so don't fail the publish attempt, just ignore it.

@dagood, where are you encountering this scenario?  In the past, effort has been made to ensure that scenario simply does not occur.

Reply:
That's correct, what changed is that there's a specific scenario in product construction where this is the clearest solution. (I'll copy this to the core-eng issue, too--I thought there was context there but there really wasn't, sorry.)

### Now

When we hit a breaking change, e.g. in CoreCLR breaking CoreFX, prodcon fails and no publishing happens. Since this is in release/2.1, there are no official builds that devs can use to see if they fixed the break, and CI can't succeed because the break fix will actually break CI because it uses the checked-in dependencies.

### Desired

When we hit a breaking change, we want to publish what prodcon built so far. This lets devs easily validate their fix, and it can pass CI. After the fix is merged, we need to *resume* the build at the point of the break. Artifacts from these builds look like this:

| | coreclr | corefx | msbuild | cli  |
| --- | --- | --- | --- | --- |
| broken | y | | y | |
| retry after fix | y | y | y | y |

Publish is a final/post-build step that only sees the set of artifacts that were involved in/produced by a prodcon build. Without some significant new logic in PipeBuild, it has no way to tell if an artifact is new to a certain build, or if it was already published as part of a broken build.

This change fixes that by allowing final publish to just run on all artifacts without tracking which things have been uploaded before.

This is also a pragmatic fix so we can get this scenario running--we *could* track which artifacts have already been published, but we'd have to work where that's stored and what corner cases it has, which is much riskier IMO.

Reply:
Ok, now I understand the reason behind the change.  Thanks!